### ISSUE TITLE
Upgrade uikit to latest version

### ISSUE THREAD

**Issue by walterrowe** [glFusion/glfusion#412]

**Comment by leegarner** [comment]
Some preliminary work done here: https://github.com/leegarner-glfusion/uikit3. Seems do-able, and one good way to clean up HTML in the PHP code.

Included the uk-icon CSS from UIkit v2 since the SVG icons are very limited.

**Comment by walterrowe** [comment]
Do I only need to drop this uikit3 folder in my layout folder and choose it as my theme? Perhaps I should read the README.md, LOL.

Walter
--
WalterRowe.com
Columbia, Maryland
202-355-4123

**Comment by leegarner** [comment]
That should be it. Just download the source code and unzip into a directory
under public_html/layout like "uikit3", then set it as the theme. I created
a new admin user and enabled user themes, then set the theme for that user
so I can still log into my regular admin account and fix broken things or
at least check what they should look like.

I'm using glFusion 2.0 as the basis, and where I create new templates I'm
creating PR's to glFusion to get them into CMS. The UIkit login window was
one big one.

**Comment by walterrowe** [comment]
Has the UIKit 3 migration tool been helpful? What can we do to help?

Walter
--
WalterRowe.com
Columbia, Maryland
202-355-4123

**Comment by leegarner** [comment]
It's been somewhat helpful. It's still showing an issue about the nabar
container that I haven't quite figured out yet, maybe it's because I have 2
containers with varying visibility. And of course warnings about uk-icon,
but that's expected since I'm continuing to use v2 icons.

If you want make some updates, maybe fork
https://github.com/leegarner-glfusion/uikit3 and push updates there and I
can merge. I need to figure out a way to track what was done, maybe a
checklist by function e.g. "command and control ... done".

**Comment by leegarner** [comment]
Just a note, plugin-supplied templates may be an issue, but so far it seems
like deprecated classes intermix just fine. One thing is if a plugin (or
core) calls UIkit javascript it may need to check what uikit version it's
using. This works for where I have a notification popup, but it seems a
little kludgy

                if (typeof UIkit.notify === 'function') {
                    // uikit v2 theme
                    UIkit.notify('{lang_copy_success}',{timeout: 300});
                } else if (typeof UIkit.notification === 'function') {
                    // uikit v3 theme
                    UIkit.notification({
                        message: '<span uk-icon="icon: check"></span>
{lang_copy_success}',
                        timeout: 1500,
                        status: 'success',
                    });
                }

I was also thinking of using the $_SYSTEM['framework'] variable, setting it
to "uikit3" to tell the difference. The last approach I can think of is to
maintain 2 sets of theme files, but based on doing that when uikit was
introduced to glFusion it's not as trivial as it sounds.

**Comment by walterrowe** [comment]
While it is a bigger hill to climb, it would be great to convert all plugins to uikit3.

Version 2.0 will be a big release. If all provided and popular plugins are uikit3 ready at release, that would smooth the transition for people who depend on them. Those who have made custom themes will have to spend time upgrading their themes to uikit3. Any documentation we can write based on this exercise would help them get there faster. How to test and debug, what to look for, etc, would be ideal.

**Comment by leegarner** [comment]
I don't want to delay glFusion v2 for uikit3 as this will take a little
time and v2 is (I think) nearly ready for release. I think adding uikit3
support is a good idea to keep current but it doesn't add a whole lot of
benefit right away. I wouldn't consider a wholesale cutover in any case and
think that the uikit3 theme that can be used as a base theme, like cms is
now. This does bring up a couple of thoughts...

There may be places where UIkit JS functionls are called, likt the
notification popup (Forum uses this as well as several of my plugins. One
option is to wrap common javascript functions into a "utility" JS file in
the theme directory so glFusion.notify() can be called by a plugin and will
do the right thing. Here's an example to just display the notification
popup in CMS. UIkit3 would just change the call inside notify():
/* Display a popup notification */
var glFusion = (function() {
        return {
                notify: function(message, status='success') {
 ...
                        $.UIkit.notify(icon + message, {timeout:
1000,pos:'top-center'});
                }
        };
})();
... then
$outputHandle->addScriptFile($_CONF['path_layout'].'js/glfusion.js'); in
the theme's functions.php.

Also, CMS is used as the base template dir if a derived theme doesn't have
a requested template. What about adding a variable in functions.php
indicating the base theme name, that could be used by the template class to
find undefined templates? Default is "cms" for backward compatibility (but
I haven't found yet where the template class gets "cms")

**Comment by mark0263** [comment]
The template class uses the _template_set_root($root) function to define all the places where  it will look for templates. One of the last paths set is the cms/ theme path.

You can override the _template_set_root call using the global:

$TEMPLATE_OPTIONS['hook']['set_root'] = '_template_set_root';

So a theme could, in their functions.php set their own custom _template_set_root() function.  

I think this will do what you are looking for.

**Comment by mark0263** [comment]
All of this is defined in lib-cache.php.

**Comment by leegarner** [comment]
Now that was easy to find ;)  So here's what I'm thinking, let me know if
you see any value to it.  A theme can define $_SYSTEM['base_theme'] to
indicate its parent. For backward compatibility, if that's not defined it
defaults to "cms". If it is defined, and empty, then it is the base theme.

 function _template_set_root($root) {
-    global $_CONF, $_USER;
+    global $_CONF, $_USER, $_SYSTEM;

     $retval = array();

@@ -53,6 +53,17 @@ function _template_set_root($root) {
         $root = array($root);
     }

+    $base_theme = '';
+    if (isset($_SYSTEM['base_theme'])) {
+        if (!empty($_SYSTEM['base_theme'])) {
+            $base_theme = $_SYSTEM['base_theme'];
+        } else {
+            $base_theme = $_USER['theme'];  // this is the base theme
+        }
+    } else {
+        $base_theme = 'cms';
+    }
+
     foreach ($root as $r) {

         if (substr($r, -1) == '/') {
@@ -71,8 +82,8 @@ function _template_set_root($root) {
         if ( $r != '' ) {
             $retval[] = $r . '/custom';
             $retval[] = $r;
-            if ( $_USER['theme'] != 'cms' ) {
-                $retval[] = $_CONF['path_themes'] . 'cms/' .substr($r,
strlen($_CONF['path_layout']));
+            if ( $_USER['theme'] != $base_theme ) {
+                $retval[] = $_CONF['path_themes'] . $base_theme . '/'
.substr($r, strlen($_CONF['path_layout']));
             }
         }
     }
@@ -435,4 +446,4 @@ function
configmanager_select_cache_driver_helper($index = '')
     return $retval;
 }

Just pondering.

**Comment by walterrowe** [comment]
I might only change this:

$base_theme = '';
if (isset($_SYSTEM['base_theme'])) {
    if (!empty($_SYSTEM['base_theme'])) {
        $base_theme = $_SYSTEM['base_theme'];
    } else {
        $base_theme = $_USER['theme']; // this is the base theme
    }
} else {
    $base_theme = 'cms';
}

.. to this ..

$base_theme = 'cms';
if (isset($_SYSTEM['base_theme'])) {
    if (!empty($_SYSTEM['base_theme'])) {
        $base_theme = $_SYSTEM['base_theme'];
    } else {
        $base_theme = $_USER['theme']; // this is the base theme
    }
}

**Comment by mark0263** [comment]
The base theme is the theme that is guaranteed to have all the needed templates. Think of it as the 'parent' theme - where child theme can include only the customized templates needed for their theme.

I believe setting the $base_theme to the $_USER['theme'] will not accomplish the goal.  For example, if my theme is 'demo' (which is a child theme) - so $_USER['theme'] is set to 'demo', glFusion would not know to look in the 'cms' theme directory to find any missing templates from 'demo'.

I think this would be more appropriate..

```
+    $base_theme = 'cms';
+    if (isset($_SYSTEM['base_theme'])) {
+        if (!empty($_SYSTEM['base_theme'])) {
+            $base_theme = $_SYSTEM['base_theme'];
+        }
+    }
+

```

Always default to 'cms' as the base theme - if $_SYSTEM['base_theme'] is set, use it, otherwise always use 'cms'.

**Comment by leegarner** [comment]
Good point, I actually had a check for !empty($base_theme) in the loop but
removed it at the last minute while I was typing this and updated the if
block above. I also forgot about existing child themes of cms, so clearly
this works better.
